### DISCUSSION TITLE
Proposal: Target-typed `new[]` for arrays without initializers

### DISCUSSION THREAD

**Discussion by colejohnson66** [dotnet/csharplang#6587] [Language Ideas] [answered]
Currently, we have target-typed `new` support for objects. This allows the following shortening:

```diff
-SortedDictionary<ABigLongTypeNameA, ABigLongTypeNameB> myDictionary = new SortedDictionary<ABigLongTypeNameA, ABigLongTypeNameB>();
+SortedDictionary<ABigLongTypeNameA, ABigLongTypeNameB> myDictionary = new(); // much better
```

We also have target-typed `new[]` support for arrays with initializers:

```cs
// both are valid
private static readonly byte[] MAGIC = { (byte)'E', (byte)'L', (byte)'F', 0x7F };
private static readonly byte[] MAGIC = new[] { (byte)'E', (byte)'L', (byte)'F', 0x7F };
```

However, what we're missing is target-typed `new[]` support for arrays *without* initializers (ones with explicit lengths):

```cs
private static readonly byte[] MAGIC = new[4]; // I'll fill this somewhere 
```

This results in *three* compiler errors: CS0178 ("Invalid rank specifier"), CS1514 ('{' expected), and CS1513 ('}' expected).

For something as simple as the above, I can easily do `new byte[4]` to fix it, but why should I? What if I want an array of a long type name like the `SortedDictionary` above? I'm forced to write this:

```cs
private static readonly SortedDictionary<ABigLongTypeNameA, ABigLongTypeNameB>?[] _myArrayOfDictionaries =
    new SortedDictionary<ABigLongTypeNameA, ABigLongTypeNameB>?[4];
```

That's awful. Why can't I just write `= new[4]`?

**Comment by FaustVX** [comment]
I've been waiting for this for as long as the `target-typed new()` syntax was introduced.

**Comment by 333fred** [comment, marked as answer]
Duplicate of https://github.com/dotnet/csharplang/issues/2701.
